### PR TITLE
units: Introduce test-infrastructure feature

### PR DESCRIPTION
**Pull request by tcharding**
CI is red as expected because this only touches `units` but changes how the `arbitrary` feature works. Looking for a concept ACK, especially around the feature gating and module structure, also is the module useful at the crate root or should the trait definition either not be in a module or be re-exported?

Add a `test-infrastructure` feature, enabling this feature gives users stuff that is useful for writing test code.

- Create a module structure that keeps all the test infrastructure stuff together inside each module.
- Move `abitrary` stuff into the new module. This helps with code organisation and also allows users to get at test infra with or without the `arbitrary` dependency.
- Add a `TestDummy` trait (note for now only has one method)
- Implement `TestDummy` for all types in the crate.

The idea for this came out of #3341.

### CODE REVIEW

**Review by Kixunil — COMMENTED** [review]
This made me realize you're treating `units` as a top-level crate but there's no such crate. E.g. `crypto` is independent from `units`.

The trait needs to be in its own crate and depended on from other crate.

**Comment by apoelstra** [comment]
Maybe `bitcoin-arbitrary` to suggest that this new crate is (sorta) married to `arbitrary`? Or we could stick with `bitcoin-test` or something more general.

We probably should even re-export `arbitrary` itself from there to simplify our Cargo.toml declarations.

**Comment by Kixunil** [comment]
> to suggest that this new crate is (sorta) married to `arbitrary`

Is it though? I think it's an independent trait that is only tiny bit similar to `Arbitrary`.

Also now I'm thinking if it'd be better to put both the trait and impls in a bottom crate not a top one, so that top crates are not cluttered by this. The disadvantage is that `arbitrary` won't be able to call into it but maybe that idea wasn't that useful anyway.

**Comment by apoelstra** [comment]
> Also now I'm thinking if it'd be better to put both the trait and impls in a bottom crate not a top one, so that top crates are not cluttered by this.

Oh, yeah, good idea. I think this is conceptually the right thing to do.

> The disadvantage is that `arbitrary` won't be able to call into it but maybe that idea wasn't that useful anyway.

Yeah. I think it's a bit useful but probably not worth messing up the dep tree over.

**Comment by tcharding** [comment]
> This made me realize you're treating units as a top-level crate but there's no such crate.

What do you mean by 'top-level', I have a mental model where we have a simple directed graph of crates. 

> E.g. crypto is independent from units.

Currently sighash uses `Amount`

> The trait needs to be in its own crate and depended on from other crate.

Fair point, this is the same discussion we have had before about not dumping stuff in any crate just because it is depended on by all the others. Also this brings us back to the "we should provide a crate that implements templates etc. for easier testing" I don't personally want to work on that at the moment, I just want to solve the problem of "get rid of meaningless `Default` impls used in testing without rug pulling users".

@Kixunil you have me in a bind at the moment, everytime I try to do something so that we can move things to primitives you review by picking it to pieces so much that it creates ten more jobs each of those which will also be picked to pieces and it means we are grinding to a halt (as well as totally sapping my energy to work on `rust-bitcoin`). I have already agreed this week to stop doing any work on `rust-bitcoin` except stuff that directly pushes `primitives` forward but you then raise review blockers that request me to do such work. It is not possible to create `primitives`, small and perfect like you want it _and_ create a perfectly correct `rust-bitcoin` created using the normal correct development methods. We need to all get on the same page here, if the aim is get `primitives` out then lets do it, if the aim is keep chipping away at all `rust-bitcoin` crates then lets do that but we can't do both. 

`rust-bitcoin` is going to become a bit of a mess so that we can achieve our stated aim "get `primitives` done", reviews like "perhaps we should merge `TxidExt` and `WtixdExt` into a single trait" or "docs should use this sentence instead of that sentence because it is mildly better" just make us do a bunch more 24 hour iterations for improvements that could be discussed or done later. I appreciate that you like things done correctly, and I appreciate that doing things correctly is one of the main reasons why we all do free software but at the moment we have a very specific task at hand that is already difficult to do and the normal "lets do this 100% correctly or not do it at all" methodology is not going to work. I'm asking you to please put that shit on hold. @apoelstra and I have capitulated to agreeing that `primitives` is only POD, with nothing else in it, something that neither of us would do if you didn't push for it, and we are letting you run with the whole "`primitives` won't need a second version for 10 years" thing which again neither of us think is going to fly. At some stage I've got to draw a line bro and say what I already said in another post this morning - you are fucking killing me right now.

**Comment by tcharding** [comment]
Thanks for being so measured when I was obviously raging. I'm still gutted at our effort this last three weeks, I'll comment when I've cleared my head some more.

**Comment by apoelstra** [comment]
I mostly agree with the content of what @tcharding said -- though I'm much less personally bothered by it. To answer a couple of your questions:

> I'm not happy that you felt being pushed into this by me. I'd like to understand why exactly you don't like this so much. My thinking is that doing super-stupid POD `primitives` will already unlock many small projects and those that use them as a means of communication between libraries and as we progressively add more and more stuff to `primitives` we will unlock more and more downstream consumers.

If we weren't breaking stuff into POD types the movement would be much easier -- we'd basically move the types over along with their dependencies without looking too closely at the API. There would be far fewer extension traits (probably the address-related one on `Script`/`ScriptBuf` would still be necessary but not much else).

> I don't see how this strategy would be worse than waiting longer for `primitives` or breaking `primitives` often.

Once we had everything moved over we could go over the API and fix egregious problems but mostly, "if it works it works". If we just moved all the primitives types (`Transaction` and the types of its components), got it compiling and called it a day, that would be a perfectly serviceable API and we could move onto more interesting parts of the protocol.

Probably we would find rough edges, and in many cases we'd deprecate stuff and add new stuff. That's fine. Probably after a year or three we'd do a 2.0 to remove the deprecated stuff and clean up things that new rustc versions enabled to be cleaned up. In many cases we could just leave things rough and that'd be fine.

> > "`primitives` won't need a second version for 10 years" thing which again neither of us think is going to fly
> 
> There are multiple crates at 1.0.0 for several years. I don't see why ours would be special in the bad sense that it'd require frequent breakages.

Because the Bitcoin protocol is massive, quickly changing, and the entire developer ecosystem is still catching up with existing changes. (For example, almost nobody is using the Taproot extensions to Miniscript in a general way, even though Taproot was designed with these extensions in mind; there are big missing parts of PSBT needed for interactive multisignatures; there are only a couple key storage formats that allow error correction and nobody has actually implemented error correction for those; etc etc etc ... and yet meanwhile we are talking about new opcodes and BitVM and PTLCs and FROST signatures, none of which have complete implementations at all!)

**Comment by tcharding** [comment]
TL;DR I agree with what @apoelstra said.

What I think our problem is:

1. Andrew and I want to release basically what we have right now  but pulled out into `primitives` as `v1.0.0`
2. Kix has already agreed to doing this but releasing it as `v0.100.0`
3. Kix wants to release minimal POD stuff as `v1.0.0` and then slowly bring stuff over and do successive `v1.x` releases.

This all seems fine but it leaves us no way to progress because I am writing patches half thinking we are working on `v0.100.0` and half thinking we are working on `v1.0.0`, Kix seems to be reviewing as if we are working solely on `v1.0.0`. For example I try to move something over thinking "we can just fix/remove this later" but Kix is reviewing with "removing or changing anything in `primitives` breaks semver or is code churn".

Then the pressure builds up and I explode.

> though I'm much less personally bothered by it.

I am seemingly incapable of separating work from my own mental state, when I feel like we are not making progress it eats me up inside. I've currently no clue how to manage that.

**Comment by apoelstra** [comment]
>I am seemingly incapable of separating work from my own mental state, when I feel like we are not making progress it eats me up inside. I've currently no clue how to manage that.

I have two main ways I deal with this:

One is that I have several projects going at once, including a few (rust-miniscript, rust-simplicity) where there are big nontrivial problems that I can more-or-less solve at my own pace (though right now I have 30 or 40 commits queued on Sanket's review, which is mildly annoying). I also have a backlog of things like "error corection for rust-bech32" and "rust-secp context object" which I could do if I ever found myself blocked on everything else.

The other is that it never feels like we're "making no progress" here because you keep opening new PRs ;).

**Comment by tcharding** [comment]
> One is that I have several projects going at once,

Yeah, I have been trying to keep that in mind since you told me that in Nashville but this week when I thought I should try to limit things that would trigger @Kixunil to review I was left with only `rust-bitcoind-json-rpc` and `miniscript`, I have been working on the first but its boring as bat shit. And I don't want to start shaking `miniscript` wildly because of the patch set you have going, and also I don't really know what to do over there that would be useful). I guess I could work on PSBTv2. I like working or `rust-bitcoin` because its actually useful.

> The other is that it never feels like we're "making no progress" here because you keep opening new PRs ;).

Lol, I first read this and burst out laughing because I read it as "it never feels like we are making progress here because you keep opening PRs" - now that shit is funny.

But, when I re-read it correctly, opening PRs is not progress for me, merging them is. There is a a Gucci mane track that goes something like "my favourite color purple ...", no clue what it means but I always think of that song when I open notifications to see a bunch of purple.

**Comment by tcharding** [comment]
https://www.youtube.com/watch?v=fFMQR4WobsI

**Comment by Kixunil** [comment]
Thanks for explaining! Maybe it's not really that clear-cut as I've thought and I shouldn't have pushed that hard. I guess I was too excited about getting `ln-types` and possibly other crates to 1.0 soon. Sticking with the plan now makes sense though to avoid the chaos of plan change.

That being said I'm not sure if you've considered these things:

* There's a bunch of things that are tangled together quite a lot. Inherent methods on transaction types accepting or returning types from outside of its tree. Cutting them out would take quite some work. Even seemingly simple stuff like making secp256k1 optional is already quite crazy.
* I think cutting out POD methods from types and putting everything else should be mentally easier than figuring out the dependencies. It should be also easier to review. And while we'll have to do it at some point anyway, this way we can have a crate out sooner.
* We still need to review the entire API before publishing 1.0 (the Rust API guidelines checklist) and the more stuff there is the more work we have.

> quickly changing

Haha, tell that to shitcoiners... :laughing: 

It's not really changing that much in a way that'd require API change. I think SegWit was likely the only one that changed the transaction structure and address in a way nobody could predict/prepare for before it happened. I think the biggest trigger of change is in consensus validation where it takes new context with the changes. Makes me think that one really needs to be a separate crate.

> in a general way

Honestly, I think that the use cases of general miniscript are quite limited. Basically people need to be "lawyers" to write/read it, so general population will probably never use it and for programmers it'd be more useful to generate an API from it that would allow one to produce `script_pubkey` and sign inputs locked by it. I was thinking about this for a looong time but never found the time to play with the idea more.

> I am writing patches half thinking we are working on `v0.100.0` and half thinking we are working on `v1.0.0`

Would it help to think if it more like MVP/first iteration? I would myself like to release `primitives` with a little bit more stuff than just PODs (`COINBASE*` constants are a good example here), I just don't want to delay the moving by figuring out what those things should be. If you come across anything you feel like should go in `primitives` right away you can write it down somewhere and we can discuss it later or even in parallel.

>  opening PRs is not progress for me, merging them is

I have some idea about this being a known psychological phenomenon. What might possibly help is for you to make a separate checklist of smaller tasks ("address this review", "move this code"...) and check that off when done. I kinda do this with the "Done" button on GH notifications.

> I read it as "it never feels like we are making progress here because you keep opening PRs"

LOL me too. It both looked like a joke but also as an inappropriate joke in the current situation, so I was confused until I've read your response.

**Comment by apoelstra** [comment]
> That being said I'm not sure if you've considered these things:
> 
> * There's a bunch of things that are tangled together quite a lot. Inherent methods on transaction types accepting or returning types from outside of its tree. Cutting them out would take quite some work. Even seemingly simple stuff like making secp256k1 optional is already quite crazy.

For sure -- if it were straightforward we'd have done it a while ago. We did have several abortive attempts and I think without the idea to use extension traits we'd still be stuck.

> * I think cutting out POD methods from types and putting everything else should be mentally easier than figuring out the dependencies. It should be also easier to review. And while we'll have to do it at some point anyway, this way we can have a crate out sooner.

Yes, if there were a clearcut distinction between "POD methods" and the rest. Even easier than what we're doing would be "if it compiles when you move it, move it" (and if we later find it's limiting some other move, remove it) and not worrying too much beyond that.

This is just an extreme example of a process that would've worked and been faster than what we're doing. Obviously we should have some criteria beyond "compiler says this PR is ok". But if our current process makes @tcharding feel like he's just spinning his wheels then maybe we should move in that direction.

> * We still need to review the entire API before publishing 1.0 (the Rust API guidelines checklist) and the more stuff there is the more work we have.

Sure, we ought to do this, but this is really just a "bonus" for people who want a stable rust-bitcoin crate. It's just not that important if we have weirdly named or redundant methods.

> It's not really changing that much in a way that'd require API change. I think SegWit was likely the only one that changed the transaction structure and address in a way nobody could predict/prepare for before it happened...

Taproot introduced a new mostly-overlapping script system, with completely different resource limits, a new "OP_SUCCESSx" mechanism, new complicated data structures that interact with it, etc. It also came with a new address type (bech32m) that tweaked the "designed to be future-proof" bech32 format.

Silent payments are even more invasive because they change the way that addresses are used entirely. Cross-signature aggregation would be similarly disruptive. For wallets, the same applies to any form of interactive signing protocol.

I'm sure our existing APIs will hold up fine, but we're going to need to add new stuff which is partially redundant with the old stuff, and the resulting conceptual framework won't be as clean as it would be if we got to redesign the whole crate ecosystem every 5 years. But it means that there is a point at which efforts to create a perfect API are mostly wasted work. I'm not saying we're at that point now, but I am saying that this point will happen for rust-bitcoin long before it'd happen for most 1.0 crates (or even or own leaf crates like `hex-conservative` or `ordered`).

>Honestly, I think that the use cases of general miniscript are quite limited

In context, I think your point with this comment is that "of course there's limited uptake of miniscript, it doesn't have usecases". But general miniscript is very widely used and deployed. It's specifically the Taproot stuff that has limited update, despite having clear privacy and efficiency benefits, and my point is this is because it's a new thing that requires nontrivial engineering work to become usable.

And because it breaks existing concepts/assumptions) such as e.g. there being a limit on the "total size" of a miniscript) while inttroducing a new one (there being a "depth limit" of 128), there is difficult API work to integrate it, and the result won't be as nice as it would've been had we somehow started designing everything after knowing how to use Taproot.

>LOL me too. It both looked like a joke but also as an inappropriate joke in the current situation

Sorry :) I'll try to avoid double-negatives like that. I did mean it as a bit of a mean joke -- that Tobin's sysiphean work is sysiphean work that *I* don't have to do -- but it was too easy to misread as a different kind of mean joke.

**Review by liamaharon — COMMENTED** [review]
DQ, why not feature gate with `#[cfg(test)]`?

**Comment by apoelstra** [comment]
> DQ, why not feature gate all this test stuff with `#[cfg(test)]`?

Because `cfg(test)` stuff isn't available to dependencies of the crate. (And in a multi-crate repo, this means that e.g. test functionality in `units` won't be available to `primitives`.)

**Comment by Kixunil** [comment]
> if there were a clearcut distinction between "POD methods" and the rest.

I believe we have agreed to just doing required conversion methods - those that cannot be implemented by external crates because they require access to the internal fields of the type. It seems very clear to me. Completely public types don't need anything at all, types like `Script` have `from_bytes` and `as_bytes` which are pretty obvious. You could argue we broke the rule because `into_boxed_script` wasn't really required because it can be implemented using `unsafe` but that relies on `repr(transparent)` which we currently tell to not rely on and the method is super obvious conversion copied from `std`.

**Comment by apoelstra** [comment]
>True, though "too much work, we didn't optimize, let's give up" sucks.

How about "we spent months with nearly zero output but at least we didn't "give up""?

>Sounds like we have a different understanding of what "leaf crate" means. To me, the crates that almost everything else depends on are root crates, not leaf.

Good to be aware of. I think it's more intuitive to use "root" for the one that there's one of. But I think we all understand what we're referring to.

>Interesting! If you have a list somewhere I'd like to see it. Note that I'm specifically looking for cases where the user is inputting miniscript

Both anchorwatch and liana have built their businesses around this. There is an online interactive editor that dariorsor built that several people have shown me (though I can't find it now).

**Comment by tcharding** [comment]
This is going nowhere, closing.
